### PR TITLE
Assert terminator support

### PR DESCRIPTION
**Pull request by atgeller**
Assert, a terminator in rust's mir, seems to be generated for under/overflow checking, as well as checking for div-by-zero. This includes supporting it essentially as a guarded goto, but also demonstrates the code needed to validate the absence of possible assert failures.

### CODE REVIEW

File: liquid-rust-typeck/src/checker.rs [hunk 1]
```diff
@@ -302,6 +305,33 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
+        let cond_ty = self.check_operand(env, cond);
+
+        let pred = match cond_ty.kind() {
+            TyKind::Refine(BaseTy::Bool, e) => {
+                e.clone()
+            }
+            _ => unreachable!("unexpected cond_ty {:?}", cond_ty),
+        };
+
+        let assert = ExprKind::BinaryOp(BinOp::Eq, pred, ExprKind::Constant(liquid_rust_fixpoint::Constant::Bool(*expected)).intern()).intern();
```
Comment (nilehmann) [inline liquid-rust-typeck/src/checker.rs:325]: I would go with the following as I think it'll generate a slightly simpler constraint:

```suggestion
        let assert = if cond { pred } else { pred.not() }
```

File: liquid-rust-typeck/src/checker.rs [hunk 2]
```diff
@@ -302,6 +305,33 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
         Ok(())
     }
 
+    fn check_assert(
+        &mut self,
+        env: &mut TypeEnv<'tcx>,
+        cursor: &mut Cursor,
+        cond: &Operand,
+        expected: &bool,
```
Comment (nilehmann) [inline liquid-rust-typeck/src/checker.rs:313]:
```suggestion
        expected: bool,
```
https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

File: tests/neg/assert_terminator.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#![feature(register_tool)]
+#![register_tool(lr)]
+
+// Note: only fails if line 329 of liquid-rust-typeck/src/checker.rs is uncommented
+#[lr::ty(fn(usize, usize) -> usize)]
+pub fn assert_terminator_test(a: usize, b: usize) -> usize {
```
Comment (nilehmann) [inline tests/neg/assert_terminator.rs:6]: You need to add test explicitly to `liquid-rust-driver/tests/pos.rs`

File: tests/pos/assert_terminator.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#![feature(register_tool)]
+#![register_tool(lr)]
+
+#[lr::ty(fn(usize, usize{x: x > 0}) -> usize)]
+pub fn assert_terminator_test(a: usize, b: usize) -> usize {
```
Comment (nilehmann) [inline tests/pos/assert_terminator.rs:5]: You need to add test explicitly to `liquid-rust-driver/tests/neg.rs`

File: liquid-rust-typeck/src/checker.rs [hunk 5]
```diff
@@ -302,6 +305,33 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
+            TyKind::Refine(BaseTy::Bool, e) => {
+                e.clone()
+            }
+            _ => unreachable!("unexpected cond_ty {:?}", cond_ty),
+        };
+
+        let assert = ExprKind::BinaryOp(BinOp::Eq, pred, ExprKind::Constant(liquid_rust_fixpoint::Constant::Bool(*expected)).intern()).intern();
+
+        // Uncomment the below line to allow pre-catching of possible divide-by-zero, underflow, and overflow
+        // WARNING: rust is very eager about inserting under/overflow checks, so be warned that uncommenting this will likely break everything
```
Comment (nilehmann) [inline liquid-rust-typeck/src/checker.rs:328]: More than a `warning` I would add a `TODO` to make this user-configurable. We have some support for config flags here https://github.com/liquid-rust/liquid-rust/blob/main/liquid-rust-common/src/config.rs so we could make a new config to choose between both. It doesn't have to go in this PR though.

File: liquid-rust-core/src/ir.rs [hunk 6]
```diff
@@ -192,6 +197,9 @@ impl fmt::Debug for Terminator {
             TerminatorKind::Drop { place, target } => {
                 write!(f, "drop({:?}) -> {:?}", place, target)
             }
+            TerminatorKind::Assert { cond, target, expected } => {
+                write!(f, "Assert({:?} is expected to be {:?}) -> {:?}", cond, expected, target)
```
Comment (nilehmann) [inline liquid-rust-core/src/ir.rs:201]:
```suggestion
                write!(f, "assert({:?} is expected to be {:?}) -> {:?}", cond, expected, target)
```

**Review by nilehmann — CHANGES_REQUESTED** [review]
@atgeller I just realized that no matter which optimization flags are passed to `rustc` it will always generate an assert for integer division and reminder, so we should land this before adding support for `%`.

Some comments below

**Comment by atgeller** [comment]
I think the latest commits have resolved everything. I'll take a look at adding an option that just generates a warning/todo if set to true for now, and we can worry about threading it through later.

**Review by nilehmann — APPROVED** [review]
Lgtm

**Pull request merged**